File: test/core/alarm/rearm_rule_test.dart
import 'package:critalarm/core/alarm/rearm_rule.dart';
import 'package:critalarm/core/push/incident_push.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  final now = DateTime.utc(2026, 9, 22, 3);
  final ringUntil = now.add(const Duration(minutes: 20));

  bool ask({
    String incidentId = 'inc_9a8b7c',
    IncidentPushKind kind = IncidentPushKind.repeat,
    bool criticalOn = true,
    bool ackedLocally = false,
    DateTime? until,
    bool quietHoursHold = false,
  }) => RearmRule.canRearm(
    incidentId: incidentId,
    kind: kind,
    criticalOn: criticalOn,
    ackedLocally: ackedLocally,
    ringUntil: until ?? ringUntil,
    now: now,
    quietHoursHold: quietHoursHold,
  );

  group('RearmRule.canRearm', () {
    test('all five inputs good lets the phone set its own next ring', () {
      expect(ask(), isTrue);
    });

    test('an acked incident does not ring again', () {
      expect(ask(ackedLocally: true), isFalse);
    });

    test('a ring_until in the past stops the loop', () {
      expect(ask(until: now.subtract(const Duration(seconds: 1))), isFalse);
    });

    test('ring_until exactly now stops the loop', () {
      expect(ask(until: now), isFalse);
    });

    test('no ring_until means the server never sent this incident', () {
      expect(
        RearmRule.canRearm(
          incidentId: 'inc_demo',
          kind: IncidentPushKind.open,
          criticalOn: true,
          ackedLocally: false,
          ringUntil: null,
          now: now,
          quietHoursHold: false,
        ),
        isFalse,
      );
    });

    test('a topic whose critical switch is off does not ring', () {
      expect(ask(criticalOn: false), isFalse);
    });

    test('quiet hours holds the ring', () {
      expect(ask(quietHoursHold: true), isFalse);
    });

    test('an empty incident id never rings', () {
      expect(ask(incidentId: ''), isFalse);
    });

    test('open, repeat and reopen all re-arm', () {
      for (final kind in [
        IncidentPushKind.open,
        IncidentPushKind.repeat,
        IncidentPushKind.reopen,
      ]) {
        expect(ask(kind: kind), isTrue, reason: kind.name);
      }
    });

    test('a forward or a state change never re-arms', () {
      for (final kind in [
        IncidentPushKind.p4,
        IncidentPushKind.p5,
        IncidentPushKind.ack,
        IncidentPushKind.close,
        IncidentPushKind.expire,
      ]) {
        expect(ask(kind: kind), isFalse, reason: kind.name);
      }
    });
  });

  group('RearmRule.nextRingAt', () {
    test('sets the next ring one repeat interval out', () {
      expect(
        RearmRule.nextRingAt(now: now, repeatIntervalS: 30),
        now.add(const Duration(seconds: 30)),
      );
    });

    test('never lands after ring_until', () {
      expect(
        RearmRule.nextRingAt(
          now: now,
          repeatIntervalS: 30,
          ringUntil: now.add(const Duration(seconds: 10)),
        ),
        isNull,
      );
    });

    test('a nonsense interval falls back to the server default', () {
      expect(
        RearmRule.nextRingAt(now: now, repeatIntervalS: 0),
        now.add(const Duration(seconds: RearmRule.defaultRepeatIntervalS)),
      );
    });
  });
}

File: android/app/src/test/kotlin/app/critalarm/alarm/RearmRuleTest.kt
package app.critalarm.alarm

import app.critalarm.push.IncidentPushKind
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class RearmRuleTest {
    private val now = 1_757_462_400_000L
    private val ringUntil = now + 20 * 60 * 1000L

    private fun ask(
        incidentId: String = "inc_9a8b7c",
        kind: IncidentPushKind = IncidentPushKind.REPEAT,
        criticalOn: Boolean = true,
        ackedLocally: Boolean = false,
        ringUntilMillis: Long? = ringUntil,
        quietHoursHold: Boolean = false,
    ) = RearmRule.canRearm(
        incidentId = incidentId,
        kind = kind,
        criticalOn = criticalOn,
        ackedLocally = ackedLocally,
        ringUntilMillis = ringUntilMillis,
        nowMillis = now,
        quietHoursHold = quietHoursHold,
    )

    @Test
    fun `all five inputs good lets the phone set its own next ring`() {
        assertTrue(ask())
    }

    @Test
    fun `an acked incident does not ring again`() {
        assertFalse(ask(ackedLocally = true))
    }

    @Test
    fun `a ring_until in the past stops the loop`() {
        assertFalse(ask(ringUntilMillis = now - 1))
    }

    @Test
    fun `ring_until exactly now stops the loop`() {
        assertFalse(ask(ringUntilMillis = now))
    }

    @Test
    fun `no ring_until means the server never sent this incident`() {
        assertFalse(ask(incidentId = "inc_demo", ringUntilMillis = null))
    }

    @Test
    fun `a topic whose critical switch is off does not ring`() {
        assertFalse(ask(criticalOn = false))
    }

    @Test
    fun `quiet hours holds the ring`() {
        assertTrue(ask(quietHoursHold = false))
        assertFalse(ask(quietHoursHold = true))
    }

    @Test
    fun `an empty incident id never rings`() {
        assertFalse(ask(incidentId = ""))
    }

    @Test
    fun `open repeat and reopen all re-arm`() {
        for (kind in listOf(IncidentPushKind.OPEN, IncidentPushKind.REPEAT, IncidentPushKind.REOPEN)) {
            assertTrue(kind.wireValue, ask(kind = kind))
        }
    }

    @Test
    fun `a forward or a state change never re-arms`() {
        val never = listOf(
            IncidentPushKind.P4,
            IncidentPushKind.P5,
            IncidentPushKind.ACK,
            IncidentPushKind.CLOSE,
            IncidentPushKind.EXPIRE,
        )
        for (kind in never) {
            assertFalse(kind.wireValue, ask(kind = kind))
        }
    }

    @Test
    fun `the next ring lands one repeat interval out`() {
        assertEquals(now + 30_000L, RearmRule.nextRingAtMillis(now, 30, ringUntil))
    }

    @Test
    fun `the next ring never lands after ring_until`() {
        assertNull(RearmRule.nextRingAtMillis(now, 30, now + 10_000L))
    }

    @Test
    fun `a nonsense interval falls back to the server default`() {
        assertEquals(
            now + RearmRule.DEFAULT_REPEAT_INTERVAL_S * 1000L,
            RearmRule.nextRingAtMillis(now, 0, ringUntil),
        )
    }
}
